en/lessons/basics/collections.md

```diff
@@ -3,120 +3,224 @@ version: 1.3.1
 This means that accessing the list length is an operation that will run in linear time (`O(n)`).
 For this reason, it is typically faster to prepend than to append:
 
-```elixir
-iex> list = [3.14, :pie, "Apple"]
-[3.14, :pie, "Apple"]
+```Haskell
+iex> list = ["Orange", "Banana", "Apple"]
+["Orange", "Banana", "Apple"]
 # Prepending (fast)
-iex> ["π" | list]
-["π", 3.14, :pie, "Apple"]
+iex> "Grape" : list
+["Grape", "Orange", "Banana", "Apple"]
 # Appending (slow)
 iex> list ++ ["Cherry"]
-[3.14, :pie, "Apple", "Cherry"]
+["Orange", "Banana", "Apple", "Cherry"]
 ```
 
+Lists are also a first introduction to an "inductive" or "recursive" type. Here is an example that is identical to the haskell implementation without the syntactic sugar (we use backticks to add infix syntax).
+
+```haskell
+data List a = Nil | a `Cons` List a
+```
+
+You can see that this type is recursive. `Nil` is the base case, and the `Cons` constructor joins an `a` and a recursive call to `List a`. You can also see why lists can only contain 1 type, because the `a` is threaded through the entire structure. In our definition we can replace `Nil` with the empty list `[]` and `Cons` with `:` to get back to the builtin syntax.
+
+In that case our original example breaks down like so:
+
+```haskell
+Prelude> [1,2,3,4]
+[1,2,3,4]
+Prelude> 1 : 2 : 3 : 4 : []
+[1,2,3,4]
+Prelude> 1 `Cons` 2 `Cons` 3 `Cons` 4 `Cons` Nil
+1 `Cons` (2 `Cons` (3 `Cons` (4 `Cons` Nil)))
+```
+__Note__: If you want to get the `List` example working in ghci you will have to input some special commands to make it work ``data List a = Nil | a `Cons` List a deriving Show; infixr 5 `Cons`;``.
+
+
 ### List Concatenation
 
-List concatenation uses the `++/2` operator:
+List concatenation uses the `++` operator:
 
-```elixir
-iex> [1, 2] ++ [3, 4, 1]
+```haskell
+Prelude> [1, 2] ++ [3, 4, 1]
 [1, 2, 3, 4, 1]
 ```
 
-A side note about the name (`++/2`) format used above:
-In Elixir (and Erlang, upon which Elixir is built), a function or operator name has two components: the name you give it (here `++`) and its _arity_.
-Arity is a core part of speaking about Elixir (and Erlang) code.
-It is the number of arguments a given function takes (two, in this case).
-Arity and the given name are combined with a slash. We'll talk more about this later; this knowledge will help you understand the notation for now.
+A side note about another operator that you might come across `<>`, this is an example of typeclass abstraction in haskell, it allows us to associate specific implementations of functions (and operators) with specific types. In the case of lists `<>` is exactly `++` under the hood. The full typeclass implemetation looks like this:
+
+```haskell
+instance Semigroup [a] where
+  (<>) = (++)
+```
+
+The keywords aren't important, but it should give you an intuition for when you see `<>` which just means concatenation for a specific type!
 
-### List Subtraction
+### List Difference
```

Comment:
in my opinion this section can be skipped :) users should use a better fitting data type such as `Set` for such use cases anyway.

Reply:
Interesting, I kind of agree, but left it to keep semi-parity with the elixir example. Would be interested to hear other's thoughts on it. I also thought about going the other way and showing all the set operations on lists, but that seemed like unnecessary complexity.

Reply:
Yeah it makes a lot of sense to keep semi-parity with the elixir guides. In this case though I'd say that it isn't a good idea -  because lists are such a convenient data structure we unfortunately tend to use them in places where they are inappropriate, so at least I suggest not demonstrating "bad practices" (those that will lead to incorrect results or bad performance) in tutorials even if Elixir do it :slightly_smiling_face:
And on the plus side it's less work :smiley:

Reply:
Btw this is why I would also suggest to leave out association lists in favor of `Map` :)

Reply:
Okay, you've convinced me!